ops/testing.py

```diff
@@ -2884,7 +2890,8 @@ def secret_revoke(self, id: str, relation_id: int, *, unit: Optional[str] = None
 
     def secret_remove(self, id: str, *, revision: Optional[int] = None) -> None:
         secret = self._ensure_secret(id)
-        self._ensure_secret_owner(secret)
+        if not self._has_secret_owner_permission(secret):
+            raise RuntimeError(f'You must own secret {secret.id!r} to perform this operation')
```

Comment:
I'm not a big fan of [ab]using RuntimeError, but if that's what happens in `ops`, then Harness should raise exactly same, as as similar exception as it can.

Reply:
P.S. a non-breaking change would be to have ops raise and exception that inherits from RuntimeError because it's documented as "Raised when an error is detected that doesn’t fall in any of the other categories. The associated value is a string indicating what precisely went wrong."

Reply:
Side note, just for the record, I'm not a fan of how Ops (ab)uses `RuntimeError` either -- have often wondered why and thought that was probably a bad idea. Usually (I think?) it refers to the *Python* runtime.

Reply:
I've asked the general public for clarification at https://discuss.python.org/t/meaning-of-runtimeerror/56206

As I used to think that this exception type meant "big error with [Python] runtime", while some uses, like here in ops assume that it means "stuff that happens at run time" (perhaps).

Reply:
I think it's fine for user code to raise `RuntimeError` (and [I'm not alone](https://github.com/search?q=%22raise+RuntimeError%22+language%3APython+&type=code)), and I think the docs would explicitly say if it wasn't meant to be used in this way. I think it's "something has gone wrong and I haven't made a custom exception class for this thing", and that it implies that generally it should not be caught. ISTM that's what the docs is saying, although I agree with Dima's post that it's a bit quippy.

> I'm not a big fan of [ab]using RuntimeError, but if that's what happens in `ops`, then Harness should raise exactly same, as as similar exception as it can.

This is not what happens in ops. What happens in ops is that everything is fine, there's no exception, and then when ops has finished and Juju is wrapping up the event, it complains and puts the unit in an error state. Harness doesn't have a way to model that, and I think we definitely do want to raise some sort of error rather than let it pass without complaint just because the error when running in Juju happens after ops.main finishes.

That said, I have no objection to changing this so that there's a (name to be decided) `JujuError` that gets raised instead. It's used a *lot* not just in Harness but in ops, but if our preference is to move away from that, then I think it's fine to make this specific and leave everything else alone (the raising here is new, and also not meant to be caught, so backwards compatibility isn't an issue).

Reply:
Charm-Tech discussed this today, and we decided that we would prefer to have custom exception classes rather than `RuntimeError`, but that we would prefer to do that universally rather than incrementally, and that it's a low-priority improvement for now, so can be dealt with another day.

For this PR, we're ok with keeping `RuntimeError`.

Reply:
Charm-Tech discussed this today, and we decided that we would prefer to have custom exception classes rather than `RuntimeError`, but that we would prefer to do that universally rather than incrementally, and that it's a low-priority improvement for now, so can be dealt with another day.

For this PR, we're ok with keeping `RuntimeError`.